### ISSUE TITLE
type hints and mypy

### ISSUE THREAD

**Issue by obestwalter** [tox-dev/tox#831]
@gaborbernat and me recently discussed about introducing type hints. I stumbled over pyannotate to help adding them, if we decide to do so and started to add some boilerplate in a branch. I won't have the time to pursue that, but if anyone wants to experiment with it - this would be the starting point: https://github.com/tox-dev/tox/tree/pyannotate-type-hints

**Comment by RonnyPfannschmidt** [comment]
@obestwalter for ease of taking a look please consider opening a [wip] pr

**Comment by obestwalter** [comment]
ah .. o.k. will do :)

**Comment by obestwalter** [comment]
Hi @gaborbernat, you mentioned doing something similar in a different branch with a different tool. I just had a look and have a few questions, because I am not quite sure if we are talking about the same thing. I just started to look into the whole topic, so I might have some gaps and misconceptions ...

From glimpsing at the stubs, they seem to be meant to exist outside of the code base like PyCharm did it with skeletons. The "new" way of doing that is to have a type shed then for libraries. They would live in something like https://github.com/python/typeshed for the stdlib. IIUC they would not be part of the source code. The intention for these stubs is mainly to provide type hints for IDEs to improve code completion behaviour.

The approach I was having in mind would contain the type information right in the source code, e.g.

```python
def greet_world(): # type: () -> None
    print "Hello, world!"

[...]

def call(self, *args, **kwargs):
         # type: (*str, **str) -> str
         request = make_request(*args, **kwargs)
         return self.do_api_query(request)
``` 
(from http://mypy.readthedocs.io/en/latest/cheat_sheet.html)

Or am I missing something and those stubs are also used by a tool (or mypy directly) to automatically add the type hints to the code base?

**Comment by obestwalter** [comment]
Thanks for the clarification. I just checked https://github.com/python/mypy/issues/5028 and saw that you are driving that :+1: - I will have a look at the blog artice also.

**Comment by obestwalter** [comment]
Here's a thought: once we dropped Python2.7 we can start applying [gradual typing](https://www.python.org/dev/peps/pep-0483/#summary-of-gradual-typing) manually where it makes sense (e.g. where it helps humans and IDEs to figure out the more dynamic corners in tox). 

I don't really think that anything else is needed. tox is not a library where others need type hints in type sheds or wherever so we should keep it simple here.

Mechanically adding type hints to the whole code base also feels wrong to me.

**Comment by gaborbernat** [comment]
I personally would prefer it beforehand. it's blocked for now on merging stubs and sources at mypy level. At tox is a library for our plugin developers.

**Comment by gaborbernat** [comment]
We'll do this after https://github.com/tox-dev/tox/issues/1130.

**Comment by jugmac00** [comment]
I'd say this issue could be closed - analogous to #1130 - as the `tox` rewrite branch uses typing.
